scripts/technical-debt-metrics.sh

```diff
@@ -0,0 +1,101 @@
+  "backwards compatibility flags"  "set_option.*backward"
+  "skipAssignedInstances flags"    "set_option tactic.skipAssignedInstances"
+  "adaptation notes"               "adaptation_note"
+  "disabled simpNF lints"          "nolint simpNF"
+  "disabled deprecation lints"     "set_option linter.deprecated false"
+  "erw"                            "erw \["
+)
+
+printf '|Current number|Change|Type|\n|-:|:-:|:-|\n'
+for i in ${!titlesAndRegexes[@]}; do
+  # loop on the odd-indexed entries and name each entry and the following
+  if (( (i + 1) % 2 )); then
+    title="${titlesAndRegexes[$i]}"
+    regex="${titlesAndRegexes[$(( i + 1 ))]}"
+    if [ "${title}" == "porting notes" ]
+    then fl="-i"  # just for porting notes we ignore the case in the regex
+    else fl="--"
+    fi
+    printf '%s|%s\n' "$(git grep "${fl}" "${regex}" | wc -l)" "${title}"
+  fi
+done
+
+printf '%s|%s\n' "$(grep -c 'docBlame' scripts/nolints.json)" "documentation nolint entries"
+printf '%s|%s\n' "$(grep -c 'ERR_MOD' scripts/style-exceptions.txt)" "missing module docstrings"
+printf '%s|%s\n' "$(grep -c 'ERR_NUM_LIN' scripts/style-exceptions.txt)" "large files"
+# We print the number of files, not the number of matches --- hence, the nested grep.
+printf '%s|%s\n' "$(git grep -c 'autoImplicit true' | grep -c -v 'test')" "non-test files with autoImplicit true"
+#printf '%s|%s\n' "$(git grep '@\[.*deprecated' | grep -c -v 'deprecated .*(since := "')" "deprecations without a date"
+
+initFiles="$(git ls-files '**/Init/*.lean' | xargs wc -l | sed 's=^ *==')"
+
+printf '%s|%s\n' "$(printf '%s' "${initFiles}" | wc -l)" "\`Init\` files"
+printf '%s|%s\n\n' "$(printf '%s\n' "${initFiles}" | grep total | sed 's= total==')"  'total LoC in `Init` files'
+
+printf $'```spoiler Changed \'Init\' lines by file\n%s\n```\n' "$(
+    printf '%s\n' "${initFiles}" | awk 'BEGIN{print("|LoC|Change|File|\n|-:|:-:|-|")} {printf("%s|%s\n", $1, $2)}'
+  )"
+}
+
+# collect the technical debts from the current mathlib
+new="$(git checkout -q "${currCommit}" && tdc; git switch -q -)"
+
+# collect the technical debts from the reference mathlib -- switch to the
+old="$(git checkout -q "${refCommit}" && tdc; git switch -q -)"
+
+# place the outputs side-by-side, using `@` as a separator
+paste -d@ <(echo "$new") <(echo "${old}") | sed 's=@=|@|=' |
+  # each line should look like eithe
+  # [new number]|description@[old number]|descr
+  # or something that does not start with [number]|
+  # we split the lines into "words", using `|` as a separator
+  awk -F'|' '
+    # if the first "word" is a number, then we write the 1st entry and compare it with the 4th
+    ($1+0 == $1) { printf("|%s|%s|%s|\n", $1, $1-$4, $2) }
+    # otherwise, the line is a "formatting" line, so we simply print it unchanged until we find `@`
+    !($1+0 == $1) {
+      found=0
+      for(i=1; i<=NF; i++) {
+        if ($i == "@") { found=1 }
+        if (found == "0") { printf("%s|", $i) }
+      }
```

Comment:
I don't really know awk either, but could we use a `break` statement here instead?
```suggestion
      for(i=1; i<=NF; i++) {
        if ($i == "@") break
        printf("%s|", $i)
      }
```

Reply:
Bryan thanks, this works very well!